# HG changeset patch
# User Rik <rik@octave.org>
# Date 1519667133 28800
#      Mon Feb 26 09:45:33 2018 -0800
# Node ID 1cacfebaed3e702be573dbe87fa5c34f0488682b
# Parent  1fa1869650ccd5989fe620a2d6cdb01fa2d68e3b
area.m: Implement EdgeAlpha and FaceAlpha properties (bug #53214).

* area.m: Use addproperty() to add EdgeAlpha and FaceAlpha to hggroup
properties.  Use addlistener to connect hggroup properties to update_prop()
callback.
* area.m (update_prop): Rename from update_props.  Rewrite to accept
an additional input which is the property to update.  Update only the
specified property rather than all of them for performance.

diff --git a/scripts/plot/draw/area.m b/scripts/plot/draw/area.m
--- a/scripts/plot/draw/area.m
+++ b/scripts/plot/draw/area.m
@@ -163,47 +163,46 @@ function retval = __area__ (ax, x, y, bv
                      lc, "parent", hg);
     endif
 
     y0 = y1;
 
     addproperty ("basevalue", hg, "data", bv);
     addlistener (hg, "basevalue", @move_baseline);
 
+    addproperty ("edgealpha", hg, "patchedgealpha", get (h, "edgealpha"));
     addproperty ("edgecolor", hg, "patchedgecolor", get (h, "edgecolor"));
+    addproperty ("facealpha", hg, "patchfacealpha", get (h, "facealpha"));
     addproperty ("facecolor", hg, "patchfacecolor", get (h, "facecolor"));
     addproperty ("linestyle", hg, "patchlinestyle", get (h, "linestyle"));
     addproperty ("linewidth", hg, "patchlinewidth", get (h, "linewidth"));
 
-    addlistener (hg, "edgecolor", @update_props);
-    addlistener (hg, "facecolor", @update_props);
-    addlistener (hg, "linestyle", @update_props);
-    addlistener (hg, "linewidth", @update_props);
+    addlistener (hg, "edgealpha", {@update_prop, "edgealpha"});
+    addlistener (hg, "edgecolor", {@update_prop, "edgecolor"});
+    addlistener (hg, "facealpha", {@update_prop, "facealpha"});
+    addlistener (hg, "facecolor", {@update_prop, "facecolor"});
+    addlistener (hg, "linestyle", {@update_prop, "linestyle"});
+    addlistener (hg, "linewidth", {@update_prop, "linewidth"});
 
     addproperty ("areagroup", hg, "data");
     set (retval, "areagroup", retval);
 
-    ## Matlab property, although Octave does not implement it.
+    ## Deprecated Matlab property which Octave does not implement.
     addproperty ("hittestarea", hg, "radio", "on|{off}", "off");
 
     if (! isempty (args))
       set (hg, args{:});
     endif
   endfor
 
 endfunction
 
-function update_props (h, d)
-
+function update_prop (h, ~, prop)
   kids = get (h, "children");
-  set (kids, "edgecolor", get (h, "edgecolor"),
-             "facecolor", get (h, "facecolor"),
-             "linestyle", get (h, "linestyle"),
-             "linewidth", get (h, "linewidth"));
-
+  set (kids, prop, get (h, prop));
 endfunction
 
 function move_baseline (h, d)
   persistent recursion = false;
 
   ## Don't allow recursion
   if (! recursion)
     unwind_protect
